### PR TITLE
add rxdeep

### PR DESCRIPTION
**Pull request by loreanvictor**

### CODE REVIEW

**Comment by dai-shi** [comment]
Hi, this project is not trying to list as many projects as possible (BTW, [that project](https://github.com/dai-shi/lets-compare-global-state-with-react-hooks) tries to list many projects), but I'm happy to help with suggestions.

File: src/rxdeep/index.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { State } from 'rxdeep';
+import { useObservable } from 'rxjs-hooks';
+
+import { createApp, initialState } from '../common';
+
+const state = new State(initialState.count);
+const useCount = () => useObservable(() => state) || initialState.count;
+const useIncrement = () => () => { state.value += 1; };
```
Comment (dai-shi) [inline src/rxdeep/index.js:8]: We also need to take care of `state.dummy`. Please try using `reducer` somehow.

Comment (loreanvictor) [reply src/rxdeep/index.js:8]: [just did](https://github.com/loreanvictor/will-this-react-global-state-work-in-concurrent-mode/commit/6f8d5963f7d3b9ecf6f45f8f0ee5ec4bb8233299). shouldn't GitHub automatically pull new commits?

**Comment by loreanvictor** [comment]
> Hi, this project is not trying to list as many projects as possible (BTW, that project tries to list many projects), but I'm happy to help with suggestions.

Hi, sorry wasn't aware. Someone suggested that I should run these checks on RxDeep and I thought of creating a pull-request afterwards. Should I close this PR?

P.s. I also incorporated the dummy variable and the reducer function as you suggested, though at least ATM I don't see the changes being reflected on the code being shown on the PR (which is weird).

**Comment by dai-shi** [comment]
The change looks good.

Sorry, I didn't mean to discourage you. It's very nice to see someone using this tool. It's just that I can't maintain everything. It's actually OK and encouraged to open a PR.

Actually, let me merge this in. It's good to compare several rx libs. This is basically testing rxjs-hooks. As I understand, rxdeep has nothing to do with React and nothing wrong in concurrent mode. We might change this to name `rxjs-hooks` in the future.

**Comment by loreanvictor** [comment]
> Sorry, I didn't mean to discourage you. It's very nice to see someone using this tool. It's just that I can't maintain everything. It's actually OK and encouraged to open a PR.

I fully understand. Thanks for the nice tool anyways!

> Actually, let me merge this in. It's good to compare several rx libs. This is basically testing rxjs-hooks. As I understand, rxdeep has nothing to do with React and nothing wrong in concurrent mode. We might change this to name rxjs-hooks in the future.

I am not 100% sure. I thought the same thing, so I also tried using `re-rxjs` and `useSubscription` instead of `rxjs-hooks` for connecting RxDeep to React, and got the same results (which also differ from for example pure `re-rxjs` results). I need to investigate further to see why, but for example internally RxDeep makes inline changes to avoid duplicating large state-trees and multicasts emissions, which might be the reason why it has tearing on the last 3 checks.

**Comment by dai-shi** [comment]
I see. If that's the case, I misunderstood it. Yeah, please have a deeper look. Unfortunately, I'm not super familiar with the internal of rxjs.

Failing check 4 is critical. That shouldn't happen. (I might have merged too early...)

**Comment by loreanvictor** [comment]
> I see. If that's the case, I misunderstood it. Yeah, please have a deeper look. Unfortunately, I'm not super familiar with the internal of rxjs.
> 
> Failing check 4 is critical. That shouldn't happen. (I might have merged too early...)

yes I was checking again and I am unable to replicate 4 & 6 (need to double check to see if everything is cleaned up, I was testing with reducing multicasting by having a shared sub-state alongside a shared state).

also for rxjs-hooks, it seems to only fail at 7.

**Comment by loreanvictor** [comment]
so I double checked. evidently the 4 & 8 were caused by not using `reducer()` for mutating the state (and apparently I forgot to run `build:rxdeep` before running jest after making changes).

I created a [new PR](https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-mode/pull/21), fixing RxDeep results and adding rxjs-hooks.

I also tried multicasting on rxjs-hooks, multicasting before `selector()` actually causes further tearing, mutlicasting afterwards doesn't change anything.

**Pull request merged**